src/variableDiagnosticsProvider.ts

```diff
@@ -0,0 +1,107 @@
+'use strict';
+
+import { workspace, languages, Diagnostic, DiagnosticSeverity, DiagnosticCollection, TextDocument, Range, Position, Disposable } from 'vscode';
+
+import { event } from "./events/requestVariableEvent";
+import { VariableProcessor } from "./variableProcessor";
+
+export class VariableDiagnosticsProvider {
+
+    private diagnosticCollection: DiagnosticCollection;
+
+    public activate(subscriptions: Disposable[]) {
+        this.diagnosticCollection = languages.createDiagnosticCollection();
+
+        workspace.onDidOpenTextDocument(this.checkVariables, this, subscriptions);
+        workspace.onDidCloseTextDocument((textDocument) => {
+            this.diagnosticCollection.delete(textDocument.uri);
+        }, null, subscriptions);
+        workspace.onDidSaveTextDocument(this.checkVariables, this, subscriptions);
+
+        // Check all open documents
+        this.checkVariablesInAllTextDocuments();
+
+        event(() => {
+            this.checkVariablesInAllTextDocuments();
+        });
+    }
+
+    public dispose(): void {
+        this.diagnosticCollection.clear();
+        this.diagnosticCollection.dispose();
+    }
+
+    public async checkVariablesInAllTextDocuments() {
+        workspace.textDocuments.forEach(this.checkVariables, this);
+    }
+
+    private async checkVariables(document: TextDocument) {
+        if (document.languageId !== 'http') {
+            return;
+        }
+
+        let diagnostics: Diagnostic[] = [];
+
+        let vars = this.findVariables(document);
+
+        let varNames = vars.map((v) => v.variableName.split(".")[0].split("[")[0]);
+
+        // Distinct varNames
+        varNames = Array.from(new Set(varNames));
+
+        let existArray = await VariableProcessor.checkVariableDefinitionExists(document, varNames);
+
+        existArray.forEach((ea) => {
+            if (!ea.exists) {
+                vars.forEach((v) => {
+                    if (v.variableName === ea.name) {
+                        diagnostics.push({
+                            severity: DiagnosticSeverity.Error,
+                            range: new Range(new Position(v.lineNumber, v.startIndex), new Position(v.lineNumber, v.endIndex)),
+                            message: `${v.variableName} is not loaded in memory`,
+                            source: 'ex',
+                            code: "10",
+                        });
+                    }
+                });
+            }
+        });
+
+        this.diagnosticCollection.set(document.uri, diagnostics);
+    }
+
+    private findVariables(document: TextDocument) : Variable[] {
+        let vars: Variable[] = [];
+        let lines = document.getText().split(/\r?\n/g);
+        let pattern = /\{\{(\w+)(\.\w+|\[\d+\])*\}\}/;
```

Comment:
This seems too strict(e.g., `-`, `_` not allowed), and not consistent with `RequestVariableDefinitionRegex` in `constants.ts`, I think `\S` is OK. Do you think so?

Reply:
Well, `\S` allows `}`. Which makes for example `GET {{baseUrl}}/users/{{variableName.response.body.userId}}` difficult to match (it will match the whole thing except `GET`.

I think we should be a bit strict, possibly agree on a set of legal characters. 

We could of course write even more complex and difficult to understand regexes, but is there a lot of value in that?

Reply:
For example we probably have to allow `-` because of the header-names. But allowing any `$`, `[` or `{` etc outside of property/JSON-path lookups will make the regexes hard to use. I think we should land a consensus here before I do any more changes to regexes.

Reply:
@cbrevik I agree with your opinion towards the regex, I think the `\S` is a bit broader, let's restrict the special characters to `-`, `$`, `[`, `]`, `{`, `}` and `_`. Do you think it's OK?

Reply:
Sounds okay. I think most of this will play out in `RequestVariableCacheValueProcessor`, the specific line we're commenting now is diagnostic for **all** variables (not just request variables). I'll update and ping you back.

Reply:
@cbrevik yes, the regex to match the request variable should be broader, and the details reference path(including`.(request|response).(header|body).(path|name)`) should be parsed and extracted in the `RequestVariableCacheValueProcessor`. And the parsing result can be used both for replacing the request with actual value, and diagnostic help(tell the user the which part of path is incorrect, like first part should be request or response, second part can only ref header and body, third part should be a exist header name or a correct JSON path of JSON body).

Reply:
So we should only allow `headers` or `body` as the second part as you say?

I suppose I could narrow it down in completion and diagnostic.

Just for reference, this is the current auto-completion of `response`:
![image](https://user-images.githubusercontent.com/4932625/35957484-19a845b6-0c9c-11e8-8ca3-ae0348f99e81.png)

`request`:
![image](https://user-images.githubusercontent.com/4932625/35957544-5bdaae6a-0c9c-11e8-8677-eaf8101f2c07.png)

Reply:
@cbrevik personally I think we can narrow down to _body_ and _header_ only, I think it still would be easier to allow other options later. We just need to modify the regex and the `RequestVariableCacheValueProcessor`.